add custom pushdata

cyborg0922 opened the pull request:
- Allows fine control over OP_RETURN pushdata e.g. as discrete (variable sized) PUSHDATA elements 
e.g. allows: 6a 02 6d01 0b 68656c6c6f20776f726c64 - (changes memo.cash name to "hello world")
i.e. separate chunks: 2 bytes '6d01' then 11 bytes '68656c6c6f20776f726c64'
- Currently only allows: 6a 0f 3664303168656c6c6f20776f726c64 (i.e. 0x0f bytes as one big chunk of utf-8 encoded data). 
- This is because `len(dest).to_bytes()` is hardcoded into `construct_output_block()` so 0x0f is enforced.
- `custom_pushdata = True` simply gives total freedom over what comes after OP_RETURN. Important for things like SLP protocol (in very active development).

teran-mckinney commented on bitcash/transaction.py:
```diff
@@ -224,7 +241,11 @@ def create_p2pkh_transaction(private_key, unspents, outputs):
+    if (custom_pushdata is False):
+        output_block = construct_output_block(outputs)
+    else:
+        output_block = construct_output_block(outputs, custom_pushdata=True)
```
Why not do custom_pushdata=custom_pushdata?

teran-mckinney commented on bitcash/transaction.py:
```diff
@@ -181,11 +188,21 @@ def construct_output_block(outputs):
-            script = (OP_RETURN +
-                      len(dest).to_bytes(1, byteorder='little') +
-                      dest)
+            if (custom_pushdata is False):
```
You don't need ()'s here and below.

teran-mckinney commented:
Couple C-isms, I guess. Also, could you add tests for this?

Thanks!

excalibur0922 commented:
Yep. No worries. Give me a couple days. Got a few other things to take care
of.

On Wed, 25 Jul 2018, 02:17 Teran McKinney, <notifications@github.com> wrote:

> Couple C-isms, I guess. Also, could you add tests for this?
>
> Thanks!
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/sporestack/bitcash/pull/13#issuecomment-407422807>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHv2SEsP9ZHh-VPQgrdRdKYQYiZBvGgrks5uJyx3gaJpZM4VcNd7>
> .
>

cyborg0922 commented:
**added 4 tests to test_transaction.py and fixed 1 bug**
1 for TestSanitizeTxData
- test_message_pushdata()

3 for TestConstructOutputBlock
- test_pushdata_message
- test_long_pushdata (checks that a ValueError is raised appropriately over 220 bytes)
- test_string_pushdata (checks that a TypeError is raised appropriately with string input)

Fixed one bug
- test_long_message only makes one OP_RETURN now that MESSAGE_LIMIT increased to 220 bytes.
- just increased length of message to make it overflow into two OP_RETURNs as before

teran-mckinney commented:
Hey! I am sorry for the wait. Completely forgot about this. I've been traveling a lot.

teran-mckinney commented:
Looks good, thank you!

Please ping me, and/or send me an email if I'm behind on anything else.
